contracts/btc-staking/src/validation/full.rs

```diff
@@ -6,27 +6,65 @@ use cosmwasm_std::Binary;
 
 use {
     babylon_apis::btc_staking_api::{BTCSigType, ProofOfPossessionBtc},
+    babylon_apis::to_canonical_addr,
     babylon_btcstaking::staking::enc_verify_transaction_sig_with_output,
     babylon_schnorr_adaptor_signature::{verify_digest, AdaptorSignature},
     bitcoin::consensus::deserialize,
     cosmwasm_std::CanonicalAddr,
-    k256::schnorr::{Signature, SigningKey, VerifyingKey},
+    k256::schnorr::{Signature, VerifyingKey},
     k256::sha2::{Digest, Sha256},
 };
 
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum FullValidationError {
+    #[error("Covenant public key not found in params")]
+    MissingCovenantPublicKeyInParams,
+
+    #[error("Proof of possession is missing")]
+    MissingProofOfPossession,
+
+    #[error("Failed to parse slashing rate: {0}")]
+    InvalidSlashingRate(#[from] std::num::ParseFloatError),
+
+    #[error("Unbonding transaction must spend staking output")]
+    UnbondingTxMustSpendStakingOutput,
+
+    #[error("Invalid BTC sig type: {0}")]
+    InvalidBtcSigType(String),
+
+    #[error("The finality provider corresponding to signing key {0} is not among the staker's delegated FPs.")]
+    FpNotInDelegationList(String),
+}
+
+/// Verifies the new finality provider data.
+pub fn verify_new_fp(new_fp: &NewFinalityProvider) -> Result<(), ContractError> {
+    let fp_pk = verifying_key_from_hex(&new_fp.btc_pk_hex)?;
+
+    // FIXME: parameterise `bbn` prefix
+    let fp_address = to_canonical_addr(&new_fp.addr, "bbn")?;
+
+    let fp_pop = new_fp
+        .pop
+        .as_ref()
+        .ok_or(FullValidationError::MissingProofOfPossession)?;
+
+    verify_pop(&fp_pk, fp_address, fp_pop)?;
```

Comment:
We don't need to verify PoP actually https://github.com/babylonlabs-io/cosmos-bsn-contracts/issues/7, because BSN side does not care about the Babylon address associated with FP / BTC delegation